contracts/forecall-service/AxelarForecallService.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.9;
+import { IAxelarForecallable } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarForecallable.sol';
+import { Upgradable } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/upgradables/Upgradable.sol';
+import { IAxelarGateway } from '../interfaces/IAxelarGateway.sol';
+import { IAxelarForecallService } from '../interfaces/IAxelarForecallService.sol';
+import { IERC20 } from '../interfaces/IERC20.sol';
+
+// This should be owned by the microservice that is paying for gas.
+contract AxelarForecallService is Upgradable, IAxelarForecallService {
+    IAxelarGateway public immutable gateway;
+    address public immutable forecallOperator;
+
+    constructor(address gateway_, address forecallOperator_) {
+        gateway = IAxelarGateway(gateway_);
+        forecallOperator = forecallOperator_;
+    }
+
+    modifier onlyOperator() {
+        if (msg.sender != forecallOperator) revert NotOperator();
+
+        _;
+    }
+
+    function call(
+        bytes32 commandId,
+        string calldata sourceChain,
+        string calldata sourceAddress,
+        address contractAddress,
+        bytes calldata payload
+    ) external onlyOperator {
+        if (contractAddress == address(0)) revert InvalidContractAddress();
+
+        if (commandId == bytes32(0)) {
+            IAxelarForecallable(contractAddress).forecall(sourceChain, sourceAddress, payload);
+        } else {
+            if (gateway.isCommandExecuted(commandId)) {
+                bytes32 payloadHash = keccak256(payload);
+
+                if (gateway.isContractCallApproved(commandId, sourceChain, sourceAddress, contractAddress, payloadHash)) {
```

Comment:
You can omit this to save gas since execute will revert anyway.

Reply:
I think it's preferable to not rely on the app contract to revert